### ISSUE TITLE
Memory issue in power iteration

### ISSUE THREAD

**Issue by CheukHinHoJerry** [amirgholami/PyHessian#28]
My CUDA memory usage keeps going up after each minibatch in the power iteration. Is this expected behaviour? My understanding is that gradient of each minibatch does not carry through. I added some manual gc and cache cleaning but none of them helps.

```
After forward pass: Allocated: 4.667 GB, Max allocated: 5.404 GB, Cached: 4.947 GB
After loss.backward(): Allocated: 5.226 GB, Max allocated: 5.404 GB, Cached: 5.383 GB
After zero_grad(): Allocated: 5.226 GB, Max allocated: 5.404 GB, Cached: 5.383 GB
After Hv computation: Allocated: 4.533 GB, Max allocated: 5.826 GB, Cached: 6.105 GB
After batch cleanup: Allocated: 4.532 GB, Max allocated: 5.826 GB, Cached: 5.180 GB
Computing Hv product:  31%|████████████████████████████▍                                                              | 10/32 [00:46<00:43,  1.96s/it]After forward pass: Allocated: 5.064 GB, Max allocated: 5.826 GB, Cached: 5.411 GB
After loss.backward(): Allocated: 5.620 GB, Max allocated: 5.826 GB, Cached: 5.782 GB
After zero_grad(): Allocated: 5.620 GB, Max allocated: 5.826 GB, Cached: 5.782 GB
After Hv computation: Allocated: 4.930 GB, Max allocated: 6.220 GB, Cached: 6.476 GB
After batch cleanup: Allocated: 4.930 GB, Max allocated: 6.220 GB, Cached: 5.593 GB
Computing Hv product:  34%|███████████████████████████████▎                                                           | 11/32 [00:46<00:30,  1.46s/it]After forward pass: Allocated: 5.470 GB, Max allocated: 6.220 GB, Cached: 5.826 GB
After loss.backward(): Allocated: 6.035 GB, Max allocated: 6.220 GB, Cached: 6.222 GB
After zero_grad(): Allocated: 6.035 GB, Max allocated: 6.220 GB, Cached: 6.222 GB
After Hv computation: Allocated: 5.331 GB, Max allocated: 6.642 GB, Cached: 6.921 GB
After batch cleanup: Allocated: 5.331 GB, Max allocated: 6.642 GB, Cached: 6.021 GB
Computing Hv product:  38%|██████████████████████████████████▏                                                        | 12/32 [00:47<00:22,  1.10s/it]After forward pass: Allocated: 5.877 GB, Max allocated: 6.642 GB, Cached: 6.254 GB
After loss.backward(): Allocated: 6.445 GB, Max allocated: 6.642 GB, Cached: 6.629 GB
After zero_grad(): Allocated: 6.445 GB, Max allocated: 6.642 GB, Cached: 6.629 GB
```

**Comment by CheukHinHoJerry** [comment]
I got that fixed.

**Comment by sxbkdby8** [comment]
> I got that fixed.
Hello, I'm facing similar problems with you and it could be appreciated to share your solution, best regards.
